### PR TITLE
use stringlocalizer to validation messages that was not translated

### LINKED ISSUE #4619: IValidatableObject implementations do not use string-localized messages
The validation messages should be string localizable

### PR DESCRIPTION
**Pull request by MichaelPetrinolis**
closes #4619

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs [hunk 1]
```diff
@@ -26,8 +28,9 @@ public class CreateOpenIdApplicationViewModel : IValidatableObject
 
         public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
         {
-            return OpenIdUrlValidator.ValidateUrls(nameof(RedirectUris), RedirectUris)
-                .Union(OpenIdUrlValidator.ValidateUrls(nameof(PostLogoutRedirectUris), PostLogoutRedirectUris));
+            var S = validationContext.GetService<IStringLocalizer<CreateOpenIdApplicationViewModel>>();
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs]:
```csharp
var localizer = validationContext.GetService<IStringLocalizer<CreateOpenIdApplicationViewModel>>();

return OpenIdUrlValidator.ValidateUrls(nameof(RedirectUris), RedirectUris, localizer)
                .Union(OpenIdUrlValidator.ValidateUrls(nameof(PostLogoutRedirectUris), PostLogoutRedirectUris, localizer));
```

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/EditOpenIdApplicationViewModel.cs [hunk 2]
```diff
@@ -29,8 +31,9 @@ public class EditOpenIdApplicationViewModel : IValidatableObject
 
         public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
         {
-            return OpenIdUrlValidator.ValidateUrls(nameof(RedirectUris), RedirectUris)
-                .Union(OpenIdUrlValidator.ValidateUrls(nameof(PostLogoutRedirectUris), PostLogoutRedirectUris));
+            var S = validationContext.GetService<IStringLocalizer<EditOpenIdApplicationViewModel>>();
+            return OpenIdUrlValidator.ValidateUrls(nameof(RedirectUris), RedirectUris,S)
+                .Union(OpenIdUrlValidator.ValidateUrls(nameof(PostLogoutRedirectUris), PostLogoutRedirectUris,S));
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/EditOpenIdApplicationViewModel.cs]: Same as above, and add space before the localizer

File: src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
 using System.ComponentModel.DataAnnotations;
+using Microsoft.Extensions.DependencyInjection;
 using Microsoft.Extensions.Localization;
 
 namespace OrchardCore.OpenId.Validators
 {
     public static class OpenIdUrlValidator
     {
-        public static IEnumerable<ValidationResult> ValidateUrls(string memberName, string member, IStringLocalizer S)
+        public static IEnumerable<ValidationResult> ValidateUrls(IStringLocalizer localizer, string memberName, string member)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs]: The first change looks good

**Comment by agriffard** [comment]
Please fix the build: OpenIdUrlValidator Test is failing.

File: src/OrchardCore/OrchardCore.OpenId.Core/Services/Managers/OpenIdApplicationManager.cs [hunk 4]
```diff
@@ -17,16 +18,21 @@ namespace OrchardCore.OpenId.Services.Managers
     public class OpenIdApplicationManager<TApplication> : OpenIddictApplicationManager<TApplication>,
         IOpenIdApplicationManager where TApplication : class
     {
+        private readonly IStringLocalizer<IOpenIdApplicationManager> _localizer;
```
Comment (agriffard) [inline src/OrchardCore/OrchardCore.OpenId.Core/Services/Managers/OpenIdApplicationManager.cs]: We use `S` for a StringLocalizer.

**Comment by MichaelPetrinolis** [comment]
I fixed the failing tests and made changes according to review. A non generic IStringLocalizer is missing from the DI. All the generic string resources should be defined there and they could be used from any context.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs [hunk 5]
```diff
@@ -1,20 +1,24 @@
...
             if (member != null)
             {
+                IStringLocalizer<T> s = null;
                 foreach (var url in member.Split(new[] { ' ', ',' }, StringSplitOptions.RemoveEmptyEntries))
                 {
-                    if (!Uri.TryCreate(url, UriKind.Absolute, out Uri uri) || !uri.IsWellFormedOriginalString())
+                    if (!Uri.TryCreate(url, UriKind.Absolute, out var uri) || !uri.IsWellFormedOriginalString())
                     {
-                        yield return new ValidationResult($"{url} is not wellformed", new[] { memberName });
+                        s ??= context.GetService<IStringLocalizer<T>>();
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs]: Nothing prevent to move the declaration above

```csharp
var s ??= context.GetService<IStringLocalizer<T>>();
```

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs [hunk 6]
```diff
@@ -26,8 +26,8 @@ public class CreateOpenIdApplicationViewModel : IValidatableObject
 
         public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
         {
-            return OpenIdUrlValidator.ValidateUrls(nameof(RedirectUris), RedirectUris)
-                .Union(OpenIdUrlValidator.ValidateUrls(nameof(PostLogoutRedirectUris), PostLogoutRedirectUris));
+            return validationContext.ValidateUrls<CreateOpenIdApplicationViewModel>(nameof(RedirectUris), RedirectUris)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs]: You could use inline lambda expression

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/EditOpenIdApplicationViewModel.cs [hunk 7]
```diff
@@ -29,8 +29,8 @@ public class EditOpenIdApplicationViewModel : IValidatableObject
 
         public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
         {
-            return OpenIdUrlValidator.ValidateUrls(nameof(RedirectUris), RedirectUris)
-                .Union(OpenIdUrlValidator.ValidateUrls(nameof(PostLogoutRedirectUris), PostLogoutRedirectUris));
+            return validationContext.ValidateUrls<EditOpenIdApplicationViewModel>(nameof(RedirectUris), RedirectUris)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/EditOpenIdApplicationViewModel.cs]: Same here

File: src/OrchardCore/OrchardCore.OpenId.Core/Services/Managers/OpenIdApplicationManager.cs (hunk 4 above)
Comment (hishamco) [reply src/OrchardCore/OrchardCore.OpenId.Core/Services/Managers/OpenIdApplicationManager.cs]: We agreed to use `S` in all the places as @agriffard mentioned before

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/ApplicationControllerTests.cs [hunk 8]
```diff
@@ -119,7 +119,13 @@ public async Task RedirectUrisAreValid(string uris, bool expectValidModel)
             model.AllowAuthorizationCodeFlow = true;
             model.RedirectUris = uris;
 
-            foreach (var validation in model.Validate(new ValidationContext(model)))
+            var validationContext = new ValidationContext(model);
+            var localizerMock = new Mock<IStringLocalizer<CreateOpenIdApplicationViewModel>>();
+            localizerMock.Setup(x => x[It.IsAny<string>(), It.IsAny<object[]>()])
+                .Returns((string name, object[] args) => new LocalizedString(string.Format(name, args), string.Format(name, args)));
```
Comment (hishamco) [inline test/OrchardCore.Tests/Modules/OrchardCore.OpenId/ApplicationControllerTests.cs]: `Returns((name, args)`

Why `string.Format(name, args)` used twice for both resource name & value?

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs (hunk 6 above)
Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs]: the PR was to add translation in validation messages from OpenId. I don't see any added value changing to inline lamda expression here.

File: src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs (hunk 5 above)
Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs]: you cannot do that.

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/ApplicationControllerTests.cs (hunk 8 above)
Comment (MichaelPetrinolis) [reply test/OrchardCore.Tests/Modules/OrchardCore.OpenId/ApplicationControllerTests.cs]: Did not know exactly what resource and value meant in the context of localized string.

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs (hunk 6 above)
Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs]: > the PR was to add translation in validation messages from OpenId

Yep, but there 're many formatting fixes that you did ;)

> I don't see any added value changing to inline lamda expression here.

We usually used lambda when the function contains one line

File: test/OrchardCore.Tests/Modules/OrchardCore.OpenId/ApplicationControllerTests.cs (hunk 8 above)
Comment (hishamco) [reply test/OrchardCore.Tests/Modules/OrchardCore.OpenId/ApplicationControllerTests.cs]: The 'name` is the key that you looking for, while `value` is the returned localized string according to the specified `key`

File: src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs (hunk 6 above)
Comment (MichaelPetrinolis) [reply src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs]: although it is more readable as a function :)

Comment (hishamco) [reply src/OrchardCore.Modules/OrchardCore.OpenId/ViewModels/CreateOpenIdApplicationViewModel.cs]: The important here is the localization stuff ;)

File: src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs [hunk 9]
```diff
@@ -1,20 +1,24 @@
...
 using System.ComponentModel.DataAnnotations;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Localization;
 
 namespace OrchardCore.OpenId.Validators
 {
     public static class OpenIdUrlValidator
     {
-        public static IEnumerable<ValidationResult> ValidateUrls(string memberName, string member)
+        public static IEnumerable<ValidationResult> ValidateUrls<T>(this ValidationContext context, string memberName, string member)
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.OpenId/Validators/OpenIdUrlValidator.cs]: Could you try without accepting a generic param?

**Comment by agriffard** [comment]
Please fix the conflict.

**Comment by agriffard** [comment]
Can you please rebase the PR so there is only your changes?

**Pull request merged**